tools/generate_events.py

```diff
@@ -0,0 +1,298 @@
+#!/usr/bin/env python3
+
+import yaml
+import logging
+import argparse
+import jinja2
+from collections import OrderedDict
+
+
+EVENTS_DEFINITION = 'x-events'
+
+# NOQA
+EVENT_TEMPLATE = jinja2.Template('''// Code generated by from events definition; DO NOT EDIT.
+package {{package_name}}
+
+import (
+    "context"
+    "fmt"
+    "strings"
+    "time"
+    "github.com/openshift/assisted-service/internal/events"
+{% for import in extra_imports %}
+    {{import}}
+{% endfor -%}
+)
+
+{% for event in generator -%}
+//
+// Event {{event['name']}}
+//
+{% set eventName = event.event_class() -%}
+{% set eventBaseName = event.package_private(eventName) + "Base" -%}
+{% set baseEvent = event.base_event(event) -%}
+type {{eventName}} struct {
+{%- for p, t in event.properties.items() %}
+    {{event.pascal_case(p)}} {{event.go_type(t)}}
+{%- endfor %}
+}
+
+var {{eventName}}Name string = "{{event['name']}}"
+
+func New{{eventName}}(
+{%- for p, t in event.properties.items() %}
+    {{event.camel_case(p)}} {{event.go_type(t)}},
+{%- endfor %}
+) *{{eventName}} {
+    return &{{eventName}}{
+{%- for p, t in event.properties.items() %}
+        {{event.pascal_case(p)}}: {{event.camel_case(p)}},
+{%- endfor %}
+    }
+}
+
+func Send{{eventName}}(
+    ctx context.Context,
+    eventsHandler events.Sender,
+{%- for p, t in event.properties.items() %}
+    {{event.camel_case(p)}} {{event.go_type(t)}},
+{%- endfor %}) {
+    ev := New{{eventName}}(
+{%- for p, t in event.properties.items() %}
+        {{event.camel_case(p)}},
+{%- endfor %}
+    )
+    eventsHandler.Send{{event.pascal_case(event.type)}}Event(ctx, ev)
+}
+
+func Send{{eventName}}AtTime(
+    ctx context.Context,
+    eventsHandler events.Sender,
+{%- for p, t in event.properties.items() %}
+    {{event.camel_case(p)}} {{event.go_type(t)}},
+{%- endfor %}
+    eventTime time.Time) {
+    ev := New{{eventName}}(
+{%- for p, t in event.properties.items() %}
+        {{event.camel_case(p)}},
+{%- endfor %}
+    )
+    eventsHandler.Send{{event.pascal_case(event.type)}}EventAtTime(ctx, ev, eventTime)
+}
+
+func (e *{{eventName}}) GetName() string {
+    return "{{event['name']}}"
+}
+
+func (e *{{eventName}}) GetSeverity() string {
+    {% if event.properties.severity -%}     return e.Severity
+    {%- else -%}      return "{{event.event_severity()}}"
+    {%- endif %}
+}
+
+func (e *{{eventName}}) GetClusterId() *strfmt.UUID {
+    return &e.ClusterId
+}
+
+{%- if event.type != "cluster" %}
+func (e *{{eventName}}) GetHostId() *strfmt.UUID {
```

Comment:
Regarding the late-binding conversation above ... Maybe these should be conditional on the properties provided rather than dictated explicitly based on the event type? Then things would be easier to change if/when we need a host event without a cluster id.

Reply:
Any thoughts on this one? @filanov ?

Reply:
@carbonin thanks for bringing this option. 
The developer that adds the event should know the context of the event, and based on that context to define the type of the event.

If we'll have two types of host-events (in context of infraenv or cluster), we won't have to rely on existence of certain properties to determine the context of the event and which methods it should expose. 

That can be achieved by having the following types for host event:

- host - the common host event type (as within the current PR), will generate event that implements the HostEvent interface
  - BaseEvent (GetName, GetSeverity, FormatMessage)
  - GetHostID
  - GetClusterID

- lb_host - the late-binding host event type, will generate event that implements the LBHostEvent (open for other naming) interface:
  - BaseEvent (GetName, GetSeverity, FormatMessage)
  - GetHostID
  - GetInfraEnvID
  - GetClusterID // optional

Having an explicit type for LB will allow to validate the correctness of the properties specified for instantiating the event. E.g, if the type is lb_host, it is required to include the InfraEnvID, the same way the validation checks that host event type requires clusterID as a property (as in [validate_events](https://github.com/openshift/assisted-service/pull/1854/files#diff-47228a8d6bca4abbc484035491d0e5e8db24eafdd4083e1ad8373637132bfe95R213))

Since events are being added once, the cost of specifying is minor and the outcome is an expected event that enforces the caller to provide all of the required properties.

WDYT ? 
Also, do you see any reason for this discussion to block this PR ?

Reply:
> Also, do you see any reason for this discussion to block this PR ?

If the solution was to change the initial design, then yes. But I think adding a separate event type as a part of the late binding effort works. The naming can still be up for debate, but as long as we have an option that's just an additive change then this should be okay as is.